tests/upgrades/test_capsule.py

```diff
@@ -20,256 +20,231 @@
-        content_view.repository = [repo]
-        content_view = content_view.update(['repository'])
-        content_view.publish()
-        promote(content_view.read().version[0], ak_env.id)
-        self.assertEqual(content_view.read().environment[-1].id, ak_env.id)
-        wait_untill_capsule_sync(self.cap_host)
-        org_name = entities.Organization().search(query={'search': f'id={self.org_id}'})[0].label
-        result = execute(
-            lambda: run(
-                '[ -f /var/lib/pulp/published/yum/http/repos/'
-                '{}/{}/{}/custom/{}/{}/Packages/c/{} ]; echo $?'.format(
-                    org_name,
-                    ak_env.name,
-                    self.cv_name,
-                    self.prod_name,
-                    self.repo_name,
-                    self.rpm_name,
-                )
-            ),
-            host=self.cap_host,
-        )[self.cap_host]
-        self.assertEqual('0', result)
-
-
-class Scenario_capsule_sync_3(APITestCase):
-    """The test class contains a post-upgrade scenario to test if puppet repo added to
-    satellite is synced to External capsule.
-
-    Test Steps:
-
-        1. Upgrade Satellite and Capsule.
-        2. Sync a puppet repo, Add to C.V. in satellite.
-        3. Run capsule sync and check if it's completed successfully.
-
-    BZ: 1746589
     """
+    org = entities.Organization().search(query={'search': f'name="{DEFAULT_ORG}"'})[0]
+    pre_test_name = [
+        mark.kwargs['depend_on'].__name__
+        for mark in request.node.own_markers
+        if 'depend_on' in mark.kwargs
+    ][0]
+    rpm_name = rpm1.split('/')[-1]
+    cap_host = settings.upgrade.rhev_cap_host or settings.upgrade.capsule_hostname
+    activation_key = settings.upgrade.rhev_capsule_ak or settings.upgrade.capsule_ak
+    ak = entities.ActivationKey(organization=org.id).search(
+        query={'search': f'name={activation_key}'}
+    )[0]
+    repo = entities.Repository(organization=org.id).search(
+        query={'search': f'name={pre_test_name}_repo'}
+    )[0]
+    product = entities.Product(organization=org.id).search(
+        query={'search': f'name={pre_test_name}_prod'}
+    )[0]
+    env_name = ak.environment.read()
+    org_name = env_name.organization.read_json()['label']
+
+    content_view = entities.ContentView(organization=f'{org.id}').search(
+        query={'search': f'name={pre_test_name}_cv'}
+    )[0]
+    capsule = entities.SmartProxy().search(query={'search': f'name={cap_host}'})[0]
+    call_entity_method_with_timeout(entities.Capsule(id=capsule.id).content_sync, timeout=3600)
+    result = execute(
+        lambda: run(
+            f'[ -f /var/lib/pulp/published/yum/http/repos/'
+            f'{org_name}/{env_name.name}/{content_view.name}/custom/{pre_test_name}_prod/'
+            f'{pre_test_name}_repo/Packages/b/{rpm_name} ]; echo $?'
+        ),
+        host=cap_host,
+    )[cap_host]
+    assert result == '0'
+    __cleanup(content_view, repo, product)
```

Comment:
There is an important pattern here that should be used in pytest. 

A failure at any statement or assertion in the test function will immediately stop execution of the test function, and transition to `teardown` phase.

This means that calling `__cleanup` at the end of the test is ineffective if the test fails in any way - and we want to run cleanup regardless.

There are two approaches to dealing with this behavior.  I'd say either is acceptable and are good patterns.  One of these is easily applicable to your current test structure, the other would require more refactoring around handling the upgrade `depends_on` marker, but I'll describe both.

1.  Use a yielding fixture. Ideal behavior is the entities that need to be cleaned up are parameters of the fixture, and are provided by other fixtures, or created within the fixture function itself.  After `yield`, you can add statements to the fixture that will execute during `teardown` phase. The fixture function does not have to do anything before calling `yield` - so you can write a fixture function that only does cleanup.  This would require more refactoring for a clean pattern in your tests.

2. Use `request.addfinalizer` to register the function to be executed during `teardown` phase.  This should be called as early as possible within the test, to avoid any failures preventing cleanup from running.  `addfinalizer` takes a callable, and you can use a lambda to pass the arguments.  

The request pattern looks like: `request.addfinalizer(lambda: __cleanup(content_view, repo, product))`, and in this particular function I would add that line at L126, immediately after `content_view`, `repo`, and `product` entities are created.

Reply:
Thanks, Mike for your valuable inputs, it really helps to clean the environment even if some failure happens in mid of the execution, the second approach is more appropriate for these test cases, Similar kind of thing I included in the new pytest conversion PR https://github.com/SatelliteQE/robottelo/pull/8186.